### ISSUE TITLE
Asynchronous lifecycle `change_state`

### ISSUE THREAD

**Issue by tgroechel** [ros2/rclcpp#2213]
## Feature request
Allow for user transition callbacks within lifecycle nodes (e.g., `on_configure`) to defer work and be cancelable.

Underlying scaffolding / depends on: https://github.com/ros2/rclcpp/issues/2212 + https://github.com/ros2/rclcpp/pull/2211
Initial PR: PRASYNC

## Feature description
Lifecycle transition callbacks currently must be synchronous. With the limitation of calling service within a service https://github.com/ros2/rclcpp/pull/773 and the advent of async services https://github.com/ros2/rclcpp/pull/1709, we would like to propose adding the ability to defer work within a user transition callback (e.g., `on_configure`) and allowing these transitions to be cancelled. 

Deferral + Cancelable transitions would allow for:

1. nodes/entities with **external transition dependencies to defer work until those dependencies are met** (e.g., requests to external service(s) is/are ready/complete)
2. **attempting recovery mid-transition** instead of needing to fully shutdown + fully bring the node back to a desired state

The `change_state` process would ideally be deferrable + cancelable. This would allow for both long running functions (e.g., a long load in `on_configure`) as well as free up the executor to process new events (e.g., nodes waiting on external dependencies).

[Demo Repo](https://github.com/tgroechel/lifecycle_prac/tree/defer_cancel_user_transtion_callbacks)

> Note drop downs for images / code
<details>

<summary>Expand for in-line example code</summary>

```cpp
void on_configure_async(
    const rclcpp_lifecycle::State &,
    std::shared_ptr<rclcpp_lifecycle::ChangeStateHandler> change_state_hdl) {
  // Cancel monitoring
  transition_cancel_monitoring_timer_ = create_wall_timer(
      std::chrono::milliseconds{100}, [this, change_state_hdl]() {
        if (!change_state_hdl->is_executing()) {
          transition_cancel_monitoring_timer_.reset();
          return;
        } else if (change_state_hdl->is_canceling()) { /*handle cancel*/
          size_t num_pruned_req = client_->prune_pending_requests();
          RCLCPP_INFO(this->get_logger(), "Handle cancel: pruned %ld request",
                      num_pruned_req);
          change_state_hdl->handle_canceled(true);
          transition_cancel_monitoring_timer_.reset();
        }
      });

  // Callback for future response of getting a parameter
  auto response_received_callback =
      [logger = this->get_logger(), change_state_hdl](
          rclcpp::Client<rcl_interfaces::srv::GetParameters>::SharedFuture
              future) {
        if (change_state_hdl->is_executing()) {
          auto request_response_pair = future.get();
          RCLCPP_INFO(logger, "Received parameter response: %s",
                      request_response_pair->values[0].string_value.c_str());
          change_state_hdl->send_callback_resp(
              rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
                  CallbackReturn::SUCCESS);
        }
      };

  // Sending the request and attaching the callback
  auto request =
      std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
  request->names.push_back("param1");
  RCLCPP_INFO(this->get_logger(), "Sending async param request");
  client_->async_send_request(request, std::move(response_received_callback));
} // Off executor

void on_activate_async(
  const rclcpp_lifecycle::State & state,
  std::shared_ptr<rclcpp_lifecycle::ChangeStateHandler> change_state_hdl)
{
  // create a thread to do some work passing the change_state_hdl to the
  // thread
  std::thread t(&LifecycleTalker::defer_on_activate_work, this,
    change_state_hdl);
  t.detach();
}

void defer_on_activate_work(
  std::shared_ptr<rclcpp_lifecycle::ChangeStateHandler> change_state_hdl)
{
  int sleep_time = 3;
  RCLCPP_INFO(
    this->get_logger(),
    "on_activate() {async} is called, sleeping is separate thread "
    "for %d seconds.",
    sleep_time);
  std::this_thread::sleep_for(std::chrono::seconds{sleep_time});
  RCLCPP_INFO(
    this->get_logger(),
    "on_activate() done sleeping, returning success");
  LifecycleNode::on_activate(
    state);   // activates managed entities (i.e., lifecycle_publishers)
  change_state_hdl->send_callback_resp(
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
    CallbackReturn::SUCCESS);
}
```

</details>

https://github.com/tgroechel/lifecycle_prac/assets/15292506/55600d76-9cea-40b4-8423-76d81e7e7fbf

The "ideal" would be to turn the `ChangeState.srv` into an `Action` ([docs](https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Actions/Understanding-ROS2-Actions.html)). This would give the robust capability of accept/reject a change request, publish its status, and have the ability to cancel mid transition with user attempt to handle the cancel. However, turning `change_state` into an action would break all backward compatibility (i.e., all lifecycle nodes and clients).

**We propose an initial stage, backward compatible solution here**: PRASYNC

Exact details can be found within that PR but to summarize:
- To make current lifecycle dev as well as transitioning to an `Action` in the future easier, we re-architect `LifecycleNodeInterfaceImpl` for a cleaner separation of concerns similar to [model-view-controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller). This has been seprated out into its own issue and PR: ISSUESOC PRSOC
- Created a `ChangeStateHandler` that acts very similar to a `GoalHandle` within a ROS 2 Action. The example above should give better context for this.
  - `shared_ptr` given to the user transition callback for the user to `send_callback_resp(CallbackReturn)` whenever the transition completes
  - has an equivalent `is_cancelled()` so the user can attempt to handle a cancel mid-transition
  - when a `transition_cancel` succeeds, the state follows the respective `CallbackReturn::FAILURE` path (e.g., `CONFIGURING` -> `handle_canceled(true)` -> `UNCONFIGURED`). 

<details>

<summary>Reference for `Callback::FAILURE` paths of lifecycle nodes</summary>

![image](https://github.com/tgroechel/rclcpp/assets/15292506/b5cd3b62-cbb4-4b9c-8bf7-92d553dc8357)

</details>

We hope this issue can serve as a discussion around the ability to create asynchronous state changes, whether that be through the above suggested approach and/or a long term plan to turn them into `Actions` (or other approach).

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/deferrable-canceleable-lifecycle-transitions/32318/1
